Added Room/Level Loading

ChielBruin opened the pull request:
- Added a way to load levels/rooms from .crf files. [343929f]
  - These files contain all the tiles, entities and lights for the room.
- Added WaitingController for the waiting state. [be90315]
  - This controller shows a loaded map "50x50_MenuLevel.crf"
  - The main game can be started by pausing and un-pausing the game (This code should change when the level is created and not a dummy and is therefore also not tested).
- Added tests for all made additions.
  - The overall coverage is now at 90.0% (found using EclEmma so may be a bit different in reality).

This PR removes the need for #163. 
Closes #150, closes #163.

ChielBruin committed:
Added first version of room loading.

Added the RoomReader class that reads rooms from file.
These rooms can be used as separate rooms within the levelFactories or as
a way to read entire levels from memory.

ChielBruin committed:
Adder door support

ChielBruin committed:
Merge branch 'master' into feature-roomIO

ChielBruin committed:
Cleaned up the code

Moved code to new dedicated classes, added javadocs and fixed a bug with
loading door entities.

ChielBruin committed:
Added waitingController

This controller is for the waiting state of the game.
The controller loads the tutorial level.
Pausing and unpausing the game is for now the way to continue to the main
game.

ChielBruin committed:
Changed the pointlight in RandomLevelFactory

The pointlight is now always at a correct position.

ChielBruin committed:
Added test for move method

Added a test for move(x,y,z) in the entityTest.

ChielBruin committed:
Added tests for WaitingController

Only the initialize method is not tested as it is a temporary method.

ChielBruin committed:
Added tests for the entityReader

ChielBruin committed:
Added option to load lights with random colors.

ChielBruin committed:
Added tests for LightReader

ChielBruin committed:
Added missing javadoc

ChielBruin committed:
Added tests for TileReader

ChielBruin committed:
Added tests for the RoomLoader

ChielBruin committed:
Merge branch 'dev-sprint4' into feature-roomIO

Geweldig commented on src/main/java/nl/tudelft/contextproject/controller/WaitingController.java:
```diff
@@ -0,0 +1,59 @@
+	 * Get the file to load.
+	 * @return The file to load.
+	 */
+	@SneakyThrows(URISyntaxException.class)
```
Even though this is explained in two other files in this project I'm still going to complain about it. Please add a comment explaining why sneakythrows is a suitable solution here.
(In other words, just copy it from one of those other two files)

Geweldig commented on src/main/java/nl/tudelft/contextproject/model/Bomb.java:
```diff
@@ -69,9 +70,17 @@ public void setSpatial(Spatial spatial) {
+	}
+
+	@Override
+	public void move(float x, float y, float z) {
```
This seems to be code which shouldn't be merged with this feature. Can anyone confirm?

ChielBruin replied:
It is needed to fix failing tests that require persistent PhysicsObjects.

Geweldig replied:
Ah I see, in that case this is fine.

Geweldig commented on src/main/java/nl/tudelft/contextproject/roomIO/EntityReader.java:
```diff
@@ -0,0 +1,67 @@
+	public static void readEntities(List<Entity> entities, int entityCount, int xOffset, int yOffset, BufferedReader br) throws IOException {
+		for (int i = 0; i < entityCount; i++) {
+			String in = br.readLine();
+			if (in == null) throw new IllegalArgumentException("Empty line where some data was expected when loading entity[" + i + "].");
```
I'm amazed checkstyle doesn't complain about inline ifs.

Taeir replied:
That's because we want to allow inline ifs like these, where the first statement is a return, continue, break or throw (anything that directly exits the enclosing body)

Geweldig commented:
I have left a couple comments, most of which relating to things happening in multiple files. If you agree with me on these "complaints" please fix them in every file.

ChielBruin committed:
Additions to JavaDocs

Added explanations on why @sneakyThrows was apropriate.

Diocruel commented on src/test/resources/maps/50x50_MenuLevel.crf:
```diff
@@ -0,0 +1,17 @@
+# Test level with an entity and light
```
There are two MenuLevel maps. One in src/test/resources/maps and one in src/main/....I Presume this is because then the lines of the roomreader can be tested with the map on always the same way. Is this correct?

ChielBruin replied:
That was the idea, but after your comment I checked if it was really needed and it turned out I can remove the copy safely.

Taeir commented on src/main/java/nl/tudelft/contextproject/roomIO/EntityReader.java:
```diff
@@ -0,0 +1,67 @@
+			float posx = Float.valueOf(line[0]) + xOffset;
+			float posy = Float.valueOf(line[1]);
+			float posz = Float.valueOf(line[2]) + yOffset;
+			entities.add(getEntity(line[3], posx, posy, posz, line));				
```
Nitpicking: `Float.parseFloat` is faster.
(`Float.valueOf(String)` simply performs `return new Float(parseFloat(String));`)

Taeir replied:
Perhaps add a bit of whitespace (empty lines) for readability?

Taeir commented on src/main/java/nl/tudelft/contextproject/roomIO/RoomReader.java:
```diff
@@ -0,0 +1,68 @@
+			int width = Integer.valueOf(tmp[0]);
+			int height = Integer.valueOf(tmp[1]);
+			int entityCount = Integer.valueOf(tmp[2]);
+			int lightCount = Integer.valueOf(tmp[3]);
```
Integer.parseInt is preferable.

Taeir replied:
Line 40-43 causes FindBugs Warning about unneccesary boxing/unboxing.

Taeir commented on src/test/java/nl/tudelft/contextproject/controller/WaitingControllerTest.java:
```diff
@@ -0,0 +1,63 @@
+	 * Restores the original Main instance after all tests are done.
+	 */
+	@AfterClass
+	public static void tearDownAfterClass() {
```
Move method to top of file.

Taeir commented:
Looks really good for the most part. I placed some comments about style and such. If you agree with them, please fix them. Otherwise, please state why you disagree.

On a side note, the Float.valueOf stuff is causing FindBugs warnings

Taeir commented on src/main/java/nl/tudelft/contextproject/controller/GameController.java:
```diff
@@ -34,6 +40,28 @@ public GameController(SimpleApplication app, Level level) {
+		List<Light> lights = new ArrayList<>();
+		String[] tmp = file.getName().split("_")[0].split("x");
+		int w = Integer.valueOf(tmp[0]);
+		int h = Integer.valueOf(tmp[1]);
```
Lines 54 and 55 cause FindBugs warning for unneccesary boxing/unboxing. Use `Integer.parseInt` instead.

ChielBruin committed:
Fixed findbugs warnings

Fixed findbugs warnings, moved a method to the top and changed all
valueOf() to parseInt/parseFloat().

ChielBruin committed:
Removed duplicate file.

Taeir committed:
Merge branch 'dev-sprint4' into feature-roomIO

ChielBruin committed:
Added the option to load empty tiles

Empty Tiles can now be loaded by inserting # as the tile type.

Geweldig commented:
I have nothing to complain about. If someone else agrees this can be merged.

Diocruel commented:
I agree, the loading of empty tiles is a good addition and I see that al previous comments have been handled.
